### PR TITLE
Refactor apply links

### PR DESCRIPTION
**Pull request by venpopov**
#### Summary

- add comprehensive tests for the apply links function
- add argument validation
- remove dplyr dependency and replace it with a switch statement
- pull out repeated computations of links out of the loop for efficiency
- remove "is.numeric" check for links - as far as I could tell there is no situation in which a link would be a numeric value. *Let me know if there is a reason you included this* that I have overlooked
- remove the splitting of the formula in components - the replacement logic works on the deparsed form without splitting so that seems unnecessary. All tests work the same in either version, so simpler is better

@GidonFrischkorn the main thing for you to review is if there are any edge cases that justified the much more complex logic that was there before this refactor. If so, please add tests for those cases and I will update the code to pass them.

#### Tests

[X] Confirm that all tests passed
[X] Confirm that devtools::check() produces no errors

### CODE REVIEW

**Comment by venpopov** [comment]
PS: I think this is a good example of how we should try to do PRs in the future for the main repo as well - small and self-contained, easy to review

File: tests/testthat/test-bmmformula.R [hunk 1]
```diff
@@ -84,3 +84,71 @@ test_that('print.bmmformula works', {
...
+test_that("apply_links handles NULL links input", {
+  form <- bmf(x ~ a + c, kappa ~ 1, a ~ 1, c ~ 1)
+  links <- NULL
+  reform <- apply_links(form, links)
+  expect_equal(reform, form, ignore_formula_env = TRUE)
+})
+
+test_that("apply_links is case sensitive for link names", {
+  form <- bmf(x ~ a + c, kappa ~ 1, a ~ 1, c ~ 1)
+  links <- list(a = "LOG")
```
Comment (GidonFrischkorn) [inline tests/testthat/test-bmmformula.R:152]: For this error we could include a `tolower` statement when testing for the links. This would make the passing from links insensitive to case. What do you think?

File: R/bmmformula.R [hunk 2]
```diff
@@ -453,30 +453,35 @@ is_constant.default <- function(x) {
...
-          links[[par]] == "log" ~ paste0(" exp(", par, ") "),
-          links[[par]] == "logit" ~ paste0(" inv_logit(", par, ") "),
-          links[[par]] == "probit" ~ paste0("Phi(", par, ")"),
-          TRUE ~ par
-        )
-      }
-      par_name <- paste0("\\b", par, "\\b") # match whole word only
-      split_form[2] <- gsub(par_name, replace, split_form[2])
+  stopifnot(is_bmmformula(formula))
+  stopifnot(is.list(links) || is.null(links), all(sapply(links, is.character)))
```
Comment (GidonFrischkorn) [inline R/bmmformula.R:457]: Initially, I thought passing a numeric link would be an easy way to fix a parameter to a specific value. But I see that this might not be super intuitive. Instead, we could think about adding a separate function that fixes parameters to any value passed by the user. This can sometimes make it easier to compare certain models with each other. But I also do not think this is super urgent.

**Review by GidonFrischkorn — COMMENTED** [review]
This looks good to me. And I just had one comment to make the call of different links case-insensitive.

File: R/bmmformula.R (hunk 2 above)
Comment (venpopov) [reply R/bmmformula.R:457]: I see. That's a nice idea, but I worry it adds a lot of complicated behavior. Now, when you apply a link to a parameter, this parameter is predicted in another part of the formula and is sampled. But if you apply a numeric link with the way you describe, we would have to add checks to ensure that the parameter does not appear as a predicted parameter in other parts of the formula. It also makes the apply links formula have different semantics depending on if it is a character describing a function or a number. In the first case, which is the default now, the parameter `a` id replaced by `some_fun(a)`, where some_fun is the inverse of the specified link. But with a numeric link the entire parameter would be substituted instead with the number. There are a few other complication I can think of.

Aside from all that, we already have a way to specify parameters to be constants - bmf(a ~ b +c) + bmf(b = 3). This is simpler and part of the existing functionality. If you have a model formula `myform` in which the parameters are free, you can fit a separate model with `myform + bmf(b=3)` to fit a reduced version with a fixed parameter.

File: tests/testthat/test-bmmformula.R (hunk 1 above)
Comment (venpopov) [reply tests/testthat/test-bmmformula.R:152]: I think it could cause problems down the line. function names in R are case sensitive, so we should not try to be too clever

File: R/bmmformula.R (hunk 2 above)
Comment (GidonFrischkorn) [reply R/bmmformula.R:457]: Sure. I just wanted to clarify why there was an option to pass numeric links. But I totally see your point. So, I think we should keep it the way you implemented it.

Comment (venpopov) [reply R/bmmformula.R:457]: Yes, that was really helpful to understand what to goal was!

File: R/bmmformula.R [hunk 3]
```diff
@@ -453,30 +453,35 @@ is_constant.default <- function(x) {
...
+  if (length(links) == 0) {
+    return(formula)
+  }
+  
+  pars_with_links <- names(links)
+  replacements <- sapply(pars_with_links, function(par) {
```
Comment (GidonFrischkorn) [inline R/bmmformula.R:463]: Just as a clarification, we could add `par = tolower(par)` here to make the evaluation of the link label case-insensitive.

I think this is not super relevant, and if you prefer to keep it case-sensitive, then I am also ok with it.

File: tests/testthat/test-bmmformula.R (hunk 1 above)
Comment (GidonFrischkorn) [reply tests/testthat/test-bmmformula.R:152]: But here it is technically not about a function name, but the label of the link function. I added a proposition above to the function.

But if you would rather err on the side of caution. Then I totally, agree that we can also keep it case-sensitive.

Comment (venpopov) [reply tests/testthat/test-bmmformula.R:152]: You're right. But I still think it's adding extra functionality that is not needed and could potentially cause confusion. I would prefer to keep things simple.
